include/aws/io/uri.h

```diff
@@ -0,0 +1,122 @@
+#ifndef AWS_IO_URI_H
+#define AWS_IO_URI_H
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/common/byte_buf.h>
+
+#include <aws/io/exports.h>
+
+/**
+ * Data representing a URI. uri_str is always allocated and filled in.
+ * The other portions are merely storing offsets into uri_str.
+ */
+struct aws_uri {
+    struct aws_allocator *allocator;
+    struct aws_byte_buf uri_str;
+    struct aws_byte_cursor scheme;
+    struct aws_byte_cursor authority;
+    struct aws_byte_cursor path;
+    struct aws_byte_cursor query_string;
+    struct aws_byte_cursor request_uri;
+    struct aws_byte_cursor host_name;
+    uint16_t port;
+};
+
+/**
+ * key/value pairs for a query string.
+ */
+struct aws_uri_params {
```

Comment:
[URI RFC](https://tools.ietf.org/html/rfc3986#section-3.4) says
> query components  are **often** used to carry identifying information in the form of  "key=value" pairs

Seems like query string can be something other than key=value pairs.

C# only exposes Query as a string.

If we keep this struct around, don't end the name with a plural "s"

Reply:
I might be crazy, but I seem to recall seeing parameters that were just keys, like ?a&b=c&d&e=f

Reply:
Actually, if that's true, we should add a test for that too

Reply:
For practical use cases, C# is being worthless on this one. It's quite common to need individual args. For one, the canonicalization process needs them like this so they can be properly sorted.

Reply:
I can add support for JUST A KEY though.